src/core/executor.rs

```diff
@@ -34,33 +73,80 @@ impl Executor {
             .num_threads(num_threads)
             .thread_name(move |num| format!("{prefix}{num}"))
             .build()?;
-        Ok(Executor::ThreadPool(Arc::new(pool)))
+        Ok(Executor::ThreadPool {
+            thread_pool: Arc::new(pool),
+            concurrency_limit: None,
+        })
+    }
+
+    /// Creates an Executor that dispatches the tasks in a thread pool with a concurrency limit.
+    /// This is useful to limit the number of concurrent tasks that can be mapped by a single call to `map`.
+    /// So that a single call to `map` does not spawn more than `concurrency_limit` tasks at a time in order to not
+    /// overwhelm the pool by a single operation.
+    pub fn multi_thread_with_concurrency_limit(
+        num_threads: usize,
+        prefix: &'static str,
+        concurrency_limit: usize,
+    ) -> crate::Result<Executor> {
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .thread_name(move |num| format!("{prefix}{num}"))
+            .build()?;
+        Ok(Executor::ThreadPool {
+            thread_pool: Arc::new(pool),
+            concurrency_limit: Some(concurrency_limit),
+        })
     }
 
     /// Perform a map in the thread pool.
     ///
     /// Regardless of the executor (`SingleThread` or `ThreadPool`), panics in the task
     /// will propagate to the caller.
-    pub fn map<A, R, F>(&self, f: F, args: impl Iterator<Item = A>) -> crate::Result<Vec<R>>
-    where
+    pub fn map<
         A: Send,
         R: Send,
+        AIterator: Iterator<Item = A>,
         F: Sized + Sync + Fn(A) -> crate::Result<R>,
-    {
+    >(
+        &self,
+        f: F,
+        args: AIterator,
+    ) -> crate::Result<Vec<R>> {
         match self {
             Executor::SingleThread => args.map(f).collect::<crate::Result<_>>(),
-            Executor::ThreadPool(pool) => {
+            Executor::ThreadPool {
+                thread_pool,
+                concurrency_limit,
+            } => {
+                let limiter = match concurrency_limit {
+                    Some(limit) => {
+                        if *limit == 0 {
+                            return Err(TantivyError::InternalError(
+                                "Concurrency limit cannot be zero.".to_string(),
+                            ));
+                        }
+                        Some(Arc::new(TaskLimiter::new(*limit)))
+                    }
+                    _ => None,
+                };
```

Comment:
The error for creating the threadpool with a concurrency limit should not be deferred until it is used.